Assertion failure: sanitize some values obtained from mods (yaw, velocity, etc.)

Rogier-5 opened the pull request:
I got the following assertion failure:

```
minetest/src/util/serialize.h:278: void writeF1000(irr::u8*, irr::f32): Assertion `i >= ((float)(s32)((-0x7FFFFFFF - 1) / 1000.0f)) && i <= ((float)(s32)((0x7FFFFFFF) / 1000.0f))' failed.
```

Because a mod (older version of mobs) was setting entity yaw and velocity to `-nan`.

This patch fixes this by sanitizing the float values a mod may set for objects.

Alternatively, minetest could opt to return an error to the mod. However, as such mod bugs will be intermittent, possibly rare, and may be hard to reproduce and fix, I thought it would be better just to accept the invalid values, and sanitize them.

Optionally, I could add code to print a warning message to the log ?

raymoo commented:
I support an error or warning, so that modders know what call caused the error. If it's a warning, it should print a stack trace too.

ShadowNinja commented:
#3326

Rogier-5 commented:
With (at least ?) two PRs and three different implementations, there seems to be a genuine need for a fix...

What will be needed to get this issue finally solved and one of the implementations merged ?

I think the following requirements can be formulated:
- Catch `+/-nan` and `+/-inf`
- Catch magnitudes that are too large (the serialisation code accepts a maximum of about +/-2000000 (0x7fffffff/1000)).

Optionally:
- Force small magnitude values to 0 (the smallest possible nonzero magnitude is 0.0001)
- Limit rather large magnitudes of velocity/accelaration to a more reasonable number (e.g. +/-100)
- Limit yaw to 0..360
- Limit pitch to -90..+90

Values to be checked (for player and lua entity):
- velocity
- acceleration
- yaw
- pitch
- ???

Maybe it is best indeed to throw a lua error (except maybe for the small magnitudes, and for the yaw and pitch ranges?).

tenplus1 commented:
The serialize.h error wouldn't have been so bad if it showed the actual variable in question and which line it is.

Rogier-5 commented:
I made some modifications.

- A value error for player/object pitch, yaw, acceleration and velocity is now handled using a lua error, so that lua prints the location of the error in lua code and a stack trace.
- In order to avoid mods immediately starting to cause crashes, two levels of seriousness are distinguised:
  - values way out of range or invalid (NaN). These cause a lua error and therefore an abort.
  - values that are a bit out of range. This causes a warning message, and the value is adjusted to be in the desired range.
- I limited the allowable entity velocity and acceleration to -100 .. 100. If the absolute value exceeds 1000, a lua error is generated, else it prints a message. Feedback on these limits is appreciated.
- yaw is limited to 0 .. 360; absolute values in excess of 720 cause a lua error
- pitch is limited to -90 .. 90; absolute values in excess of 720 cause a lua error

raymoo commented:
I don't think there's a need to limit entity velocity between -100 and 100. It might not make sense in mt_game, but could be useful in a game that kept lots of chunks loaded.

At the very least, you should document everywhere in lua_api.txt where values will be clipped.

raymoo commented:
I especially don't think the limit on acceleration makes sense. Lets say velocity is limited to -100 to 100. Because acceleration is limited to -100 and 100 too, it will always take at least 1 second to reach top speed. What if you want an entity to accelerate faster than that? You would have to do the hacky thing of implementing acceleration in the entity code, which won't benefit from client-side prediction.

Rogier-5 commented:
> Maybe you can make a separate PR for limiting velocity/acceleration, since that to me is an enhancement of bigger scope than just "give better debug info for bad values".

Limiting velocity and acceleration in general is a different scope indeed.

My motivation was that if the acceleration is too large, the velocity will eventually also become too large, and that may still cause a crash. But actually, the code already limits the absolute speed to 500 * BS (=5000) elsewhere (`collisionMoveSimple`). I'll use that.

Rogier-5 commented:
The code now enforces the same maximum speed as in `collision.cpp`. The acceleration is no longer limited, except by what the serialization datatype can represent.

As the code in `content_?ao.cpp` was not taking the maximum speed into account for non-physical objects, which would also cause a crash if the acceleration was too large, and the velocity got too large as a result, I adapted that too.

The code for physical objects in `collision.cpp` was not computing the new position correctly (the code for non-physical objects was correct), so I changed that as well.

Both the non-physical and physical object code now properly take the maximum speed into account when computing *both* the new speed *and* the new position.

raymoo commented:
Thank you. I didn't know that there was already a speed limit being used.

celeron55 commented:
Has anyone been running this on a server with a lot of mods, with good results?

Megaf commented:
Needs rebase. And should be merged.

Rogier-5 commented:
Rebased to latest minetest tree

nerzhul commented on src/constants.h:
```diff
@@ -104,6 +104,8 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #define SCREENSHOT_MAX_SERIAL_TRIES 1000
 
+// Maximum speed (absolute value) for active objects.
+#define OBJECT_MAX_SPEED 500
```
please set it upper, if someone want to add laser or projectiles this could be too low

Megaf replied:
What kind of units are those? Nodes/second?

Rogier-5 replied:
@nerzul: I investigated the maximum speed, and it turns out that there is an inefficiency in the collision detection code, that makes high-velocity objects require a disproportionate amount of memory and processing time. Both can easily increase quadratically with distance traveled in a single server step; worst case the increase is cubic.

Unless the collision detection code is optimized first, setting the maximum speed much higher than 500 may cause excessive computation time and/or memory use. In a 32-bit environment using an entity speed of 2000, and only a *single* entity, I consistently get huge delays and crashes due to `std::bad_alloc`...

The collision detection code does not seem trivial to fix, so for the moment, I propose leave that to be fixed at a later time, and keep the max speed as it is.

Rogier-5 commented:
Updated and rebased.

@nerzhul: what's your opinion on leaving the max speed as is until the collision detection code is fixed ? If a speed of 2000 causes memory consumption in excess of 2GB (i.e. `std::bad_alloc` crash on 32-bit system), then IMO the limit should be well below 2000.

I tested with mobs of the following types: mobf, mobs_redo, simple mobs, nssm and cme. In each case, I created loads of mobs, and had them walk around for as long as possible, and chase me. Except as noted below, I found no issues.

- Cme does a strange thing when computing the yaw, so that it will routinely exceed 720. This PR will not accept those values: there has to be some limit, and while my code accepts and adjusts values up to 720, a mod going from below 360 to over 720 in a single step really needs to be fixed IMO.

  I did create cme issue for this.

- Reportedly (#4756), simple mobs can also cause range errors, although I didn't observe any myself.

In any case, this PR will adjust values that are just a little out of range. If they are far out of range, and it throws a lua error, then the error message is much improved, printing a lua stack trace, with the exact location of the error, including the offending value, and the range of acceptable values.

ShadowNinja commented on builtin/game/misc.lua:
```diff
@@ -166,6 +166,18 @@ function core.setting_get_pos(name)
 end
 
+function core.get_world_limits()
+	local absolute_limit = 31000
```
This isn't accurate.  The real limits vary in each dimension since they're rounded to the nearest multiple of the block size (16), but they're closer to 31912.  If this isn't correct mods will stop working correctly near the world edge, which is undesirable because some people try to go there as a challenge, and several have made it.

Rogier-5 replied:
IIRC (from studying the code some time ago, while implementing this PR), 31000 is an absolute limit. No block will be generated that is beyond that limit. Ever. Actually, no _chunk_ will be generated that countains a block that exceeds the limit, or even that contains a block that is adjacent to the limit: when generating a chunk, a volume that is 1 block larger in all directions is processed; IIRC, if just one node in that volume being processed is beyond 31000, then the chunk won't be generated.

So basicallly, between the most distant node of a world that will be generated and the limit of 31000 (or whatever it is), there should always be at least one block of non-generated world.

IIRC, with a chunk size of 5, the actual limit is about 30912 in one direction, and at most 15 or maybe 31 more in the other.

sfan5 replied:
![screenshot_20161122_201548](https://cloud.githubusercontent.com/assets/1042418/20538219/85dc731e-b0f0-11e6-8fc0-7b9937d0ea97.png)

Rogier-5 commented:
Rebased to latest minetest tree

tenplus1 commented:
Is this still being looked at ? we need this pull...

Fixer-007 commented:
Related? https://forum.minetest.net/viewtopic.php?f=6&t=16537

Zeno- commented on src/util/numeric.h:
```diff
@@ -131,6 +131,28 @@ inline bool isInArea(v3s16 p, v3s16 d)
+//    (float(9876543210) - 360 * trunc(float(9876543210) / 360)
+// is not equal to:
+//    (double(9876543210) - 360 * trunc(double(9876543210) / 360)
+// Result for float(9876543210):	-352
```
Could modulo360f() be used here instead?

Fixer-007 commented:
Bump, everyone also forgot about this one.

paramat commented on builtin/game/misc.lua:
```diff
@@ -168,6 +168,18 @@ function core.setting_get_pos(name)
+	else
+		v = vector.new(value, value, value)
+	end
+	return vector.subtract(0, v), v
```
No. Mods can create structures that extend beyond 31000, the absolute limit is the limits of s16 co-ordinates.
Also, 'mapgen limit' can be, and is sometimes, reduced, but leaves world present outside the new limit. Mods can also create structures that extend beyond the 'mapgen limit' setting.

paramat commented on doc/lua_api.txt:
```diff
@@ -2089,6 +2089,7 @@ Call these functions only at load time!
 * `minetest.setting_getbool(name)`: returns boolean or `nil`
 * `minetest.setting_get_pos(name)`: returns position or nil
 * `minetest.setting_save()`, returns `nil`, save all settings to config file
+* `minetest.get_world_limits()`: returns the min and max coordinate of the world (pos, pos)
```
Not needed anymore, see above.

paramat commented on doc/lua_api.txt:
```diff
@@ -2781,7 +2782,11 @@ This is basically a reference to a C++ `ServerActiveObject`
 * `getpos()`: returns `{x=num, y=num, z=num}`
 * `setpos(pos)`; `pos`=`{x=num, y=num, z=num}`
+    * positions that are outside of the map are not accepted.
+      The world limits can be obtained using minetest.get_world_limits()
```
Needs removing.

paramat commented on doc/lua_api.txt:
```diff
@@ -2781,7 +2782,11 @@ This is basically a reference to a C++ `ServerActiveObject`
+      The world limits can be obtained using minetest.get_world_limits()
 * `moveto(pos, continuous=false)`: interpolated move
+    * positions that are outside of the map are not accepted.
+      The world limits can be obtained using minetest.get_world_limits()
```
As above.

paramat commented on src/script/lua_api/l_object.cpp:
```diff
@@ -179,6 +180,9 @@ int ObjectRef::l_setpos(lua_State *L)
 	v3f pos = checkFloatPos(L, 2);
+	u16 limit = MYMIN(MAX_MAP_GENERATION_LIMIT,
+			g_settings->getU16("map_generation_limit"));
+	check_range_v3f(L, pos / BS, -limit, limit);
```
Limit is wrong, see my other comment.

paramat commented on src/script/lua_api/l_object.cpp:
```diff
@@ -194,6 +198,9 @@ int ObjectRef::l_moveto(lua_State *L)
 	v3f pos = checkFloatPos(L, 2);
+	u16 limit = MYMIN(MAX_MAP_GENERATION_LIMIT,
+			g_settings->getU16("map_generation_limit"));
+	check_range_v3f(L, pos / BS, -limit, limit);
```
Limit is wrong, see my other comment.

paramat commented:
Author has disappeared, needs adoption. Seems an important addition but needs some reworking on world limits.

paramat commented:
Seems high priority so keeping open.

paramat commented:
Needs adoption, improvement and rebase. Keeping open due to importance.

paramat commented:
No adoption interest shown, closing for now.
